spec/mongoid/contextual/mongo_spec.rb

```diff
@@ -3640,6 +3640,75 @@
               expect(new_order.reload.genres).to eq(["electronic"])
             end
           end
+
+          context "when operation is $pull" do
+            context "when pulling single element" do
+
+              before do
+                depeche_mode.update_attribute(:genres, ["electronic", "pop"])
+                new_order.update_attribute(:genres, ["electronic", "pop"])
+                context.update_all("$pull" => { genres: "electronic" })
+              end
+
+              it "updates the first matching document" do
+                expect(depeche_mode.reload.genres).to eq(["pop"])
+              end
+
+              it "updates the last matching document" do
+                expect(new_order.reload.genres).to eq(["pop"])
+              end
+            end
+
+            context "when pulling based on condition" do
+              before do
+                depeche_mode.update_attribute(:genres, ["electronic", "pop", "dance"])
+                new_order.update_attribute(:genres, ["electronic", "pop", "dance"])
+                context.update_all("$pull" => { genres: {'$in' => ["electronic", "pop"] }})
+              end
+
+              it "updates the first matching document" do
+                expect(depeche_mode.reload.genres).to eq(["dance"])
+              end
+
+              it "updates the last matching document" do
+                expect(new_order.reload.genres).to eq(["dance"])
+              end
+            end
+          end
+
+          context "when operation is $pop" do
+
+            before do
+              depeche_mode.update_attribute(:genres, ["pop", "electronic" ])
+            end
+
+            it "removes first element in array" do
+              context.update_all("$pop" => { genres: -1 })
+              expect(depeche_mode.reload.genres).to eq(["electronic"])
+            end
+
+            it "removes last element in array" do
+              context.update_all("$pop" => { genres: 1 })
+              expect(depeche_mode.reload.genres).to eq(["pop"])
+            end
+          end
+
+          context "when operation is $pullAll" do
+
+            before do
+              depeche_mode.update_attribute(:genres, ["pop", "electronic", "dance" ])
+              new_order.update_attribute(:genres, ["electronic", "pop", "dance"])
+              context.update_all("$pullAll" => { genres: ["pop", "electronic"] })
+            end
```

Comment:
@JohnMaguir for $pullAll it is better to prepare multiple entries in the data to be pulled from the array, just to cover it fully:
```
              depeche_mode.update_attribute(:genres, ["pop", "electronic", "dance", "pop" ])
              new_order.update_attribute(:genres, ["electronic", "pop", ''electronic","dance"])
```

Reply:
Thank you @dem updated now